src/wal/src/local_storage_impl/segment.rs

```diff
@@ -0,0 +1,790 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{
+    collections::{HashMap, VecDeque},
+    fmt::Debug,
+    fs,
+    fs::{File, OpenOptions},
+    io,
+    io::Write,
+    path::Path,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+};
+
+use bytes_ext::BytesMut;
+use codec::Encoder;
+use common_types::{table::TableId, SequenceNumber, MAX_SEQUENCE_NUMBER, MIN_SEQUENCE_NUMBER};
+use generic_error::{BoxError, GenericError};
+use macros::define_result;
+use memmap2::{MmapMut, MmapOptions};
+use runtime::Runtime;
+use snafu::{ensure, Backtrace, ResultExt, Snafu};
+
+use crate::{
+    kv_encoder::{CommonLogEncoding, CommonLogKey},
+    local_storage_impl::record_encoding::{Record, RecordEncoding},
+    log_batch::{LogEntry, LogWriteBatch},
+    manager::{
+        BatchLogIteratorAdapter, Read, ReadContext, ReadRequest, RegionId, ScanContext,
+        ScanRequest, SyncLogIterator, WalLocation, WriteContext,
+    },
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to open or create file: {}", source))]
+    FileOpen { source: io::Error },
+
+    #[snafu(display("Failed to map file to memory: {}", source))]
+    Mmap { source: io::Error },
+
+    #[snafu(display("Segment full"))]
+    SegmentFull,
+
+    #[snafu(display("Failed to append data to segment file: {}", source))]
+    SegmentAppend { source: io::Error },
+
+    #[snafu(display("Failed to flush mmap: {}", source))]
+    Flush { source: io::Error },
+
+    #[snafu(display(
+        "Attempted to read beyond segment size. Offset: {}, Size: {}, FileSize: {}",
+        offset,
+        size,
+        file_size
+    ))]
+    ReadOutOfBounds {
+        offset: u64,
+        size: u64,
+        file_size: u64,
+    },
+
+    #[snafu(display("Invalid segment header"))]
+    InvalidHeader,
+
+    #[snafu(display("Segment not open, id:{}", id))]
+    SegmentNotOpen { id: u64 },
+
+    #[snafu(display("Segment not found, id:{}", id))]
+    SegmentNotFound { id: u64 },
+
+    #[snafu(display("Unable to convert slice: {}", source))]
+    Conversion {
+        source: std::array::TryFromSliceError,
+    },
+
+    #[snafu(display("{}", source))]
+    Encoding { source: GenericError },
+
+    #[snafu(display("Invalid record: {}, backtrace:\n{}", source, backtrace))]
+    InvalidRecord {
+        source: GenericError,
+        backtrace: Backtrace,
+    },
+
+    #[snafu(display("Length mismatch: expected {} but found {}", expected, actual))]
+    LengthMismatch { expected: usize, actual: usize },
+
+    #[snafu(display("Checksum mismatch: expected {}, but got {}", expected, actual))]
+    ChecksumMismatch { expected: u32, actual: u32 },
+}
+
+define_result!(Error);
+
+const HEADER: &str = "HoraeDB WAL";
```

Comment:
Usually we define it like this
```suggestion
const SEGMENT_HEADER: &[u8] = b"HoraeDBWAL";
```